### PR TITLE
support external ssh directories as dependencies and outputs

### PR DESCRIPTION
**Pull request by efiop**

### CODE REVIEW

File: dvc/remote/base.py [hunk 1]
```diff
@@ -174,13 +381,18 @@ def save(self, path_info, checksum_info):
 
         checksum = checksum_info[self.PARAM_CHECKSUM]
         if not self.changed_cache(checksum):
+            self._checkout(path_info, checksum)
             return
 
-        to_info = self.checksum_to_path_info(checksum)
+        self._save(path_info, checksum)
```
Comment (mroutis) [inline dvc/remote/base.py]: This kind of abstraction doesn't look like the best one to me: `foo` calling `_foo`.
If I understand correctly, the reason behind doing this is to have a private method that instances can overwrite to modify the behavior without modifying the public interface.

A good refactor could be to migrate to `ABC` and use `super()` to call the parent's method: https://www.python-course.eu/python3_abstract_classes.php

If this isn't the case and I'm missing something, maybe we should improve the naming, or it is the wrong abstraction, having `save` and `_save` is not intuitive.

File: dvc/remote/base.py [hunk 2]
```diff
@@ -226,6 +439,7 @@ def ospath(self):
         return posixpath
 
     def checksum_to_path(self, checksum):
+        assert checksum
```
Comment (mroutis) [inline dvc/remote/base.py]: Maybe we can use [pytype](https://github.com/google/pytype) and remove these assertions.

File: dvc/state.py [hunk 3]
```diff
@@ -29,11 +29,18 @@ def __init__(self, dvc_version, expected, actual):
         )
 
 
-def _file_metadata_changed(actual_mtime, mtime, actual_size, size):
-    return actual_mtime != mtime or actual_size != size
+class StateBase(object):
```
Comment (mroutis) [inline dvc/state.py]: What's the reason behind having another Base/Abstract class? The only one who is using `StateBase` is `State`, so it doesn't make a lot of sense to me.

It doesn't even raise `NotImplementedError`, just passes silently.

File: dvc/remote/base.py [hunk 4]
```diff
@@ -165,6 +317,61 @@ def changed(self, path_info, checksum_info):
         logger.debug("'{}' hasn't changed.".format(path_info))
         return False
 
+    def link(self, from_info, to_info):
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/base.py]: @mroutis Take a look at RemoteLOCAL.link(). It is `copy` only for some remotes. For example, ssh will also get reflink/hardlink/symlink support later.

File: dvc/remote/base.py [hunk 5]
```diff
@@ -165,6 +317,61 @@ def changed(self, path_info, checksum_info):
...
+    def is_empty(self, path_info):
+        return False
+
+    def isfile(self, path_info):
+        raise NotImplementedError
+
+    def isdir(self, path_info):
+        return False
+
+    def walk(self, path_info):
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/base.py]: @mroutis again, `os.walk` is only relevant for RemoteLOCAL. We have other remotes too, for which definition of walk() is different.

**Comment by efiop** [comment]
Thanks for a late review @mroutis ! :slightly_smiling_face:

**Pull request merged**